CRM/Utils/Crypt.php

```diff
@@ -51,20 +51,13 @@ public static function encrypt($string) {
     if (empty($string)) {
       return $string;
     }
-
-    if (function_exists('mcrypt_module_open') &&
-      defined('CIVICRM_SITE_KEY')
-    ) {
-      $td = mcrypt_module_open(MCRYPT_RIJNDAEL_256, '', MCRYPT_MODE_ECB, '');
-      // ECB mode - iv not needed - CRM-8198
-      $iv = '00000000000000000000000000000000';
-      $ks = mcrypt_enc_get_key_size($td);
-      $key = substr(sha1(CIVICRM_SITE_KEY), 0, $ks);
-
-      mcrypt_generic_init($td, $key, $iv);
-      $string = mcrypt_generic($td, $string);
-      mcrypt_generic_deinit($td);
-      mcrypt_module_close($td);
+    if (defined('CIVICRM_SITE_KEY') && function_exists('openssl_encrypt')) {
+      $raw = defined('OPENSSL_RAW_DATA') ? OPENSSL_RAW_DATA : TRUE;
+      $key = openssl_digest(CIVICRM_SITE_KEY, 'sha256');
+      $method = 'AES-256-ECB';
+      $ivSize = 0;
+      $iv = substr($key, 0, $ivSize);
+      $string = openssl_encrypt($string, $method, $key, $raw, $iv);
```

Comment:
I've been a little bit confused about the scoping of the PR. The title/description suggests that it's a formulaic change for compatibility (replacing one library with another work-a-like), but in reality it's *also* a policy change which uses different crypto params. (AES is a variant of Rijndael, but AES-256 is [different](https://paragonie.com/blog/2015/05/if-you-re-typing-word-mcrypt-into-your-code-you-re-doing-it-wrong) from mcrypt's Rijndael-256. ) This is (part of) why the on-disk format changes and why the PR requires re-encrypting the data.

Generally, I feel there are two plays that make the most sense. Either:
 * (a)  **Drop-in replacement**: Strictly preserve the on-disk format and the crypto params, but use a different implementation (eg phpseclib/openssl).
 * (b) **Design cleanup**: Deprecate `CRM_Utils_Crypt`. Keep it around (so we can still read/write the old format), but design a different API and on-disk format. Anything that calls `CRM_Utils_Crypt` should probably call something else.
     * Why? Because `CRM_Utils_Crypt` has *several issues*, like (a) deriving the storage-crypto key from a rest-authentication key and (b) using ECB mode and (c) protecting one arbitrarily chosen field and (d) allowing the stored password to become corrupt based on runtime changes (enabling/disabling a PHP module). The only argument I can make for the status-quo is that... the flaws of `CRM_Utils_Crypt` don't matter because its use-case is narrow and obscure.
     * Tackling this is a bigger scope-of-work, though. You need to step back and ask different questions.

I've been trying to figure out how we'd write a drop-in replacement. Haven't figured out how to do it with `Crypt_Rijndael` (the obvious params didn't work), but in case it helps... here's my attempt to figure it out: https://gist.github.com/totten/2a269885e86f4515e17aad962cda1160

Reply:
Well, I guess it depends on how urgent this really is:
 * If the status-quo is that all PHP 7.1 sites have a hard break, then one can argue that it's urgent, and we might accept design-compromises to deliver quickly.
 * If the status-quo is that PHP 7.1 sites get a deprecation notice (depending on their reporting/settings/thresholds), then it's not really urgent. We can take our time and only push a change to the data-structure when we think it's Really Better.

Reply:
See https://issues.civicrm.org/jira/browse/CRM-21120 for a related issue that isn't addressed by this one, though I'd have to rewrite my PR to cover this change. Short version - the problem of changing platforms that do/don't include the crypt functions cause silent failure in decrypting.